rgw: Add IOC hook for IOC Cache

hualongfeng opened the pull request:
Use immutable object cache as the local data cache of RGW.
And it can also act as D3N l1 local datacache.

Signed-off-by: Feng Hualong <hualong.feng@intel.com>


## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

mattbenjamin commented on src/common/legacy_config_opts.h:
```diff
@@ -1524,6 +1524,8 @@ OPTION(rgw_torrent_encoding, OPT_STR)    // torrent field encoding
 OPTION(rgw_torrent_origin, OPT_STR)    // torrent origin
 OPTION(rgw_torrent_sha_unit, OPT_INT)    // torrent field piece length 512K
 
+OPTION(rgw_d3n_l1_local_datacache_plugins, OPT_STR) //choice of l1 local cache
```
I think this name could be improved;  don't need l1?; instead of "plugins" how about "backend"--clarifying what the plugin does and making it singular; you can't use more than one at once, can you?

hualongfeng replied:
Yes.  This also means that use l1 datacache to represent D3N's default local datacache, making the configuration of l1 datacache and IOC cache clearer。
And I have replaced plugins with backend.

mattbenjamin commented on src/rgw/rgw_aio.cc:
```diff
@@ -105,6 +106,31 @@ Aio::OpFunc aio_abstract(Op&& op, optional_yield y) {
+template <typename Op>
+Aio::OpFunc ioc_cache_aio_abstract(Op&& op, optional_yield y,
+                               off_t read_ofs, off_t read_len, IOChook* ioc_hook) {
+  return [op = std::move(op), y, read_ofs, read_len, ioc_hook] (Aio* aio, AioResult& r) mutable{
```
space after mutable and before brace

hualongfeng replied:
resolved

mattbenjamin commented on src/rgw/rgw_ioc_dispatch.cc:
```diff
@@ -0,0 +1,270 @@
+  return init.result.get();
+}
+
+namespace {
```
is this the anonymous namespace?  why?

hualongfeng replied:
The usage I learned from rgw_aio.cc, it puts the handler and some functions into the anonymous space.  
Removing this anonymous space has no effect on this piece of code. Do I need to remove it?

mattbenjamin commented on src/rgw/rgw_ioc_dispatch.cc:
```diff
@@ -0,0 +1,270 @@
+
+}//namespace ioc
+
+
```
extra line spaces

hualongfeng replied:
resolved

mattbenjamin commented on src/rgw/rgw_ioc_dispatch.cc:
```diff
@@ -0,0 +1,270 @@
+    delete m_cache_client;
+    auto controller_path = m_cct->_conf.template get_val<std::string>(
+            "immutable_object_cache_sock");
+    m_cache_client = new CacheClient(controller_path.c_str(), m_cct);
```
indentation?

hualongfeng replied:
Does it means this?
```c++
    delete m_cache_client;
    auto controller_path = m_cct->_conf.template get_val<std::string>(
      "immutable_object_cache_sock");
    m_cache_client = new CacheClient(controller_path.c_str(), m_cct);
```
if it is OK, resolved.

hualongfeng replied:
Oh, Sorry, I just found out that the indentation you mentioned is caused by tab。

mattbenjamin commented on src/rgw/rgw_ioc_dispatch.cc:
```diff
@@ -0,0 +1,270 @@
+                                ) {
+  if(ack->type != RBDSC_READ_REPLY) {
+    // go back to read rados
+    ldout(m_cct, 5) << "rgw_hook: IOC cache don't cache object, go to rados " << oid << dendl;
```
maybe "object is not in cache" or "object not found in cache"

hualongfeng replied:
resolved

mattbenjamin reviewed (commented):
this looks good, some comments

mkogan1 commented:
jenkins test make check

mkogan1 commented:
it's possible to test the cache with the following flow:
```
make ceph-immutable-object-cache vstart


mkdir -p /mnt/nvme/rgw_datacache/ /mnt/nvme/rgw_ioc_cache/

pkill -9 ceph-immutable-object-cache
../src/stop.sh ; sleep 2 ; rm -vr ./dev ; /usr/bin/rm -vrf /mnt/nvme/rgw_datacache/* /mnt/nvme/rgw_ioc_cache/*

MON=1 OSD=1 MDS=0 MGR=1 RGW=1 ../src/vstart.sh -n -d --bluestore \
        -o rgw_d3n_l1_local_datacache_enabled=true \
        -o rgw_d3n_l1_datacache_persistent_path=/mnt/nvme/rgw_datacache/ \
        -o rgw_d3n_l1_datacache_size=10737418240 \
        -o rgw_enable_ops_log=true \
        -o rgw_d3n_local_datacache_backend=ioc_cache \
        -o immutable_object_cache_sock=./out/immutable_object_cache_sock.asok \
        -o immutable_object_cache_max_size=20G \
        -o immutable_object_cache_path=/mnt/nvme/rgw_ioc_cache/ \
        -o debug_immutable_obj_cache=20/20


pkill -9 radosgw

./bin/ceph-immutable-object-cache -c ./ceph.conf
#^^^
./bin/radosgw -c ./ceph.conf --log-file=./out/radosgw.8000.log --admin-socket=./out/radosgw.8000.asok --pid-file=./out/radosgw.8000.pid -n client.rgw.8000 --rgw_frontends="beast port=8000" --debug_ms=0 --debug_rgw=5 --rgw_d3n_l1_local_datacache_enabled=true -f


ls -ltrh /mnt/nvme/rgw_ioc_cache
fallocate -l 8M ./8M.dat
s3cmd mb s3://bkt                 
s3cmd put ./8M.dat s3://bkt      
s3cmd get s3://bkt/8M.dat --force
ls -ltrhR /mnt/nvme/rgw_ioc_cache
/mnt/nvme/rgw_ioc_cache:                                                                                                                                                                      
total 0                                                                                                                                                                                       
drwxrwxr-x 2 mkogan mkogan 124 Dec  9 13:21 45                                                                                                                                                
                                                                                                                                                                                              
/mnt/nvme/rgw_ioc_cache/45:                                                                                                                                                                   
total 4.0M                                                                                                                                                                                    
-rw-r--r-- 1 mkogan mkogan 4.0M Dec  9 13:21 :7:18446744073709551614:4a86a94e-4f74-4c4b-a07f-0b763cd34cdb.4176.1__shadow_.Mw2LBReF2K3IIOLwhsiMCmvgyIduUjY_1  
```

Note: in addition to previsly discussed params,
 `-o immutable_object_cache_sock=./out/immutable_object_cache_sock.asok`
is necessary to run  in vstart environment.

Performance measurements on 1GB file show expected acceleration 
(corroborated with `iostat` that reads are performed from the NVME device)

```
fallocate -l 1G ./1G.dat
s3cmd mb s3://bkt                 
s3cmd put ./1G.dat s3://bkt      
s3cmd get s3://bkt/1G.dat /dev/shm/1G_get.dat --force ; rm /dev/shm/1G_get.dat
download: 's3://bkt/1G.dat' -> '/dev/shm/1G_get.dat'  [1 of 1]
 1073741824 of 1073741824   100% in   14s    69.44 MB/s  done
                                             ^^^^^^^^^^
ls -lh /mnt/nvme/rgw_ioc_cache

sync; sudo sh -c "echo 3 > /proc/sys/vm/drop_caches"
s3cmd get s3://bkt/1G.dat /dev/shm/1G_get.dat --force ; rm /dev/shm/1G_get.dat
download: 's3://bkt/1G.dat' -> '/dev/shm/1G_get.dat'  [1 of 1]
 1073741824 of 1073741824   100% in    6s   159.23 MB/s  done
                                            ^^^^^^^^^^^
```

mkogan1 commented:
jenkins test make check

mkogan1 commented:
@hualongfeng thank you for rebasing, any objection to cherry-picking this PR/commit into the D3N PR #36266 and merging both together?
(would like to re-use the `file_aio_read_abstract()` infrastructure for libaio reads in the D3N PR as well)

hualongfeng commented:
> @hualongfeng thank you for rebasing, any objection to cherry-picking this PR/commit into the D3N PR #36266 and merging both together?
> (would like to re-use the `file_aio_read_abstract()` infrastructure for libaio reads in the D3N PR as well)

@mkogan1  No, you can merge the PR into D3N PR #36266 together.

mkogan1 commented:
@hualongfeng the D3N PR #36266 has merged and created some merge conflicts, mind if I work on rebasing this PR after which you can review my changes and modify as necessary?

hualongfeng commented:
> @hualongfeng the D3N PR #36266 has merged and created some merge conflicts, mind if I work on rebasing this PR after which you can review my changes and modify as necessary?

No, you can rebase this PR. And I could review the changes.

mkogan1 commented:
> > @hualongfeng the D3N PR #36266 has merged and created some merge conflicts, mind if I work on rebasing this PR after which you can review my changes and modify as necessary?
> 
> @mkogan1 I rebased the PR, could you review it? And you can do your changes.

@hualongfeng thank you very much for rebasing, will check running s3tests, java-s3tests, ragweed <- with valgrind and long duration stability under heavy load.
following that we have to think how to integrate with the d3n teuthology testing

hualongfeng commented:
> @hualongfeng thank you very much for rebasing, will check running s3tests, java-s3tests, ragweed <- with valgrind and long duration stability under heavy load.
> following that we have to think how to integrate with the d3n teuthology testing

@mkogan1  I wrote a teuthology test yaml, could you review it?

mkogan1 commented:
@hualongfeng saw _valgrind_ issues when running _s3tests_


```
# from the build directory
pwd
/mnt/osd--host/src-git/RGW_DATACACHE/ceph--hualongfeng--new_ioc_hook/build

# # # vstart
MON=1 OSD=1 MDS=0 MGR=1 RGW=1 ../src/vstart.sh -n -d --bluestore \
    -o rgw_enable_ops_log=true  \
    -o rgw_d3n_l1_local_datacache_enabled=true  \
       -o rgw_d3n_l1_local_datacache_backend=ioc_cache \
       -o immutable_object_cache_sock=./out/immutable_object_cache.asok \
       -o immutable_object_cache_max_size=20G \
       -o immutable_object_cache_path=/mnt/nvme/rgw_ioc_cache/ \
       -o debug_immutable_obj_cache=20/20

# # # after vstart.sh ...

valgrind --vgdb=yes --trace-children=no --child-silent-after-fork=yes --num-calle
rs=10 --suppressions=../qa/valgrind.supp --xml=yes --xml-file=./valgrind_log.xml --time-stamp=yes --tool=memcheck --max-threads=1024 ./bin/radosgw -c ./ceph.conf --log-file=./out/radosgw.800
0.log --admin-socket=./out/radosgw.8000.asok --pid-file=./out/radosgw.8000.pid -n client.rgw.8000 --rgw_frontends="beast port=8000" --debug_ms=0 --debug_rgw=5 -f --rgw_d3n_l1_local_datacache
_enabled=1 --rgw_enable_ops_log=true --rgw_d3n_l1_local_datacache_backend=ioc_cache

# # # from another tab

vim ../qa/workunits/rgw/run-s3tests.sh
#S3TEST_CONF=s3tests.conf.SAMPLE virtualenv/bin/python -m nose -a '!fails_on_rgw,!lifecycle_expiration,!fails_strict_rfc2616' -v
S3TEST_CONF=s3tests.conf.SAMPLE virtualenv/bin/python -m nose -a '!fails_on_rgw,!lifecycle_expiration,!fails_strict_rfc2616,!s3select,!test_of_sts,!webidentity_test,!fails_with_subdomain' -v
:x

../qa/workunits/rgw/run-s3tests.sh
```
examples of relevant valgrind issue:

```
vim valgrind_log.xml
...
<error>
  <unique>0x3cc465</unique>
  <tid>146</tid>
  <kind>InvalidRead</kind>
  <what>Invalid read of size 8</what>
  <stack>
  ┊ <frame>
  ┊ ┊ <ip>0x6E1212C</ip>
  ┊ ┊ <obj>/mnt/osd--host/src-git/RGW_DATACACHE/ceph--hualongfeng--new_ioc_hook/build/lib/libradosgw.so.2.0.0</obj>
  ┊ ┊ <fn>D3nGetObjData::D3nGetObjData()</fn>
  ┊ ┊ <dir>/mnt/osd--host/src-git/RGW_DATACACHE/ceph--hualongfeng--new_ioc_hook/build/../src/rgw</dir>
  ┊ ┊ <file>rgw_d3n_cacherequest.h</file>
  ┊ ┊ <line>23</line>
  ┊ </frame>
  </stack>
  <auxwhat>Address 0x57e5a728 is in a rw- anonymous segment</auxwhat>
</error>
...
```

mkogan1 commented:
on another run, did not get the valgrind errors from the above post but 2 _s3tests_ tests have failed:
```
s3tests_boto3.functional.test_s3.test_atomic_read_8mb ... FAIL
s3tests_boto3.functional.test_s3.test_atomic_write_8mb ... FAIL
```

mkogan1 commented:
when starting with the `d3n` backend (`--rgw_d3n_l1_local_datacache_backend=d3n`) a segfault occurs while running _s3tests_

```
 ceph version 17.0.0-6133-g12e18d32820 (12e18d328207a3b4b858e140b21da3a9ac57f1f1) quincy (dev) 
 1: /lib64/libpthread.so.0(+0x14b20) [0x69c3b20] 
 2: (D3nDataCache::lru_remove(D3nChunkDataInfo*)+0x118) [0x54be938] 
 3: (D3nDataCache::get(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, long)+0x1e4) [0x54be0b4] 
 4: (D3nRGWDataCache<RGWRados>::get_obj_iterate_cb(DoutPrefixProvider const*, rgw_raw_obj const&, long, long, long, bool, RGWObjState*, void*)+0xce1) [0x5364731] 
 5: /mnt/osd--host/src-git/RGW_DATACACHE/ceph--hualongfeng--new_ioc_hook/build/lib/libradosgw.so.2(+0xa2dcde) [0x527acde] 
 6: (RGWRados::iterate_obj(DoutPrefixProvider const*, RGWObjectCtx&, RGWBucketInfo const&, rgw_obj const&, long, long, unsigned long, int (*)(DoutPrefixProvider const*, rgw_raw_obj const&, l
ong, long, long, bool, RGWObjState*, void*), void*, optional_yield)+0x51c) [0x527a9ec] 
 7: (RGWRados::Object::Read::iterate(DoutPrefixProvider const*, long, long, RGWGetDataCB*, optional_yield)+0x205) [0x52644d5] 
 8: (RGWGetObj::execute(optional_yield)+0xf55) [0x51dc8c5] 
 9: (rgw_process_authenticated(RGWHandler_REST*, RGWOp*&, RGWRequest*, req_state*, optional_yield, bool)+0xaef) [0x4e9c53f] 
 10: (process_request(rgw::sal::Store*, RGWREST*, RGWRequest*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, rgw::auth::StrategyRegistry const&, RGW
RestfulIO*, OpsLogSocket*, optional_yield, rgw::dmclock::Scheduler*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, std::chrono::duration<unsigned long, st
d::ratio<1l, 1000000000l> >*, int*)+0x22ef) [0x4e9eb5f] 
 11: /mnt/osd--host/src-git/RGW_DATACACHE/ceph--hualongfeng--new_ioc_hook/build/lib/libradosgw.so.2(+0x553bfd) [0x4da0bfd] 
 12: /mnt/osd--host/src-git/RGW_DATACACHE/ceph--hualongfeng--new_ioc_hook/build/lib/libradosgw.so.2(+0x552aa9) [0x4d9faa9] 
 13: make_fcontext() 
 NOTE: a copy of the executable, or `objdump -rdS <executable>` is needed to interpret this. 
```

hualongfeng commented:
@mkogan1 I am looking at this issue. But the reason has not been found yet.

hualongfeng commented:
> on another run, did not get the valgrind errors from the above post but 2 _s3tests_ tests have failed:
> 
> ```
> s3tests_boto3.functional.test_s3.test_atomic_read_8mb ... FAIL
> s3tests_boto3.functional.test_s3.test_atomic_write_8mb ... FAIL
> ```

Hi, @mkogan1  the issue I have fixed,  because the operations make `head object` be cached.

hualongfeng commented:
> when starting with the `d3n` backend (`--rgw_d3n_l1_local_datacache_backend=d3n`) a segfault occurs while running _s3tests_
> 
> ```
>  ceph version 17.0.0-6133-g12e18d32820 (12e18d328207a3b4b858e140b21da3a9ac57f1f1) quincy (dev) 
>  1: /lib64/libpthread.so.0(+0x14b20) [0x69c3b20] 
>  2: (D3nDataCache::lru_remove(D3nChunkDataInfo*)+0x118) [0x54be938] 
>  3: (D3nDataCache::get(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, long)+0x1e4) [0x54be0b4] 
>  4: (D3nRGWDataCache<RGWRados>::get_obj_iterate_cb(DoutPrefixProvider const*, rgw_raw_obj const&, long, long, long, bool, RGWObjState*, void*)+0xce1) [0x5364731] 
>  5: /mnt/osd--host/src-git/RGW_DATACACHE/ceph--hualongfeng--new_ioc_hook/build/lib/libradosgw.so.2(+0xa2dcde) [0x527acde] 
>  6: (RGWRados::iterate_obj(DoutPrefixProvider const*, RGWObjectCtx&, RGWBucketInfo const&, rgw_obj const&, long, long, unsigned long, int (*)(DoutPrefixProvider const*, rgw_raw_obj const&, l
> ong, long, long, bool, RGWObjState*, void*), void*, optional_yield)+0x51c) [0x527a9ec] 
>  7: (RGWRados::Object::Read::iterate(DoutPrefixProvider const*, long, long, RGWGetDataCB*, optional_yield)+0x205) [0x52644d5] 
>  8: (RGWGetObj::execute(optional_yield)+0xf55) [0x51dc8c5] 
>  9: (rgw_process_authenticated(RGWHandler_REST*, RGWOp*&, RGWRequest*, req_state*, optional_yield, bool)+0xaef) [0x4e9c53f] 
>  10: (process_request(rgw::sal::Store*, RGWREST*, RGWRequest*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, rgw::auth::StrategyRegistry const&, RGW
> RestfulIO*, OpsLogSocket*, optional_yield, rgw::dmclock::Scheduler*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, std::chrono::duration<unsigned long, st
> d::ratio<1l, 1000000000l> >*, int*)+0x22ef) [0x4e9eb5f] 
>  11: /mnt/osd--host/src-git/RGW_DATACACHE/ceph--hualongfeng--new_ioc_hook/build/lib/libradosgw.so.2(+0x553bfd) [0x4da0bfd] 
>  12: /mnt/osd--host/src-git/RGW_DATACACHE/ceph--hualongfeng--new_ioc_hook/build/lib/libradosgw.so.2(+0x552aa9) [0x4d9faa9] 
>  13: make_fcontext() 
>  NOTE: a copy of the executable, or `objdump -rdS <executable>` is needed to interpret this. 
> ```

Hi, @mkogan1   I can't reproduce it. I use d3n cache, but is doesn't occur segfault.

hualongfeng commented:
jenkins test make check

hualongfeng commented:
jenkins test make check (arm64)

mkogan1 commented:
jenkins test make check

hualongfeng commented:
jenkins test make check arm64

hualongfeng commented:
> how would we go about submitting the teuthology test run?
> (if already done a teuthology run, can you please paste the `teuthology-suite ...` command that you are testing with)
> and related question, in the `qa/suites/rgw/verify/datacache/rgw-ioc-datacache.yaml`

Hi, @mkogan1   about teuthology run,  the teuthology environment is fixing, because I only have ubuntu environment, so I need to change centos to ubuntu.

And I will run the test, the below command will be used.

```
teuthology-suite -vv  \
                 --machine-type real \
                 --teuthology-branch master \
                 --email hualong.feng@intel.com \
                 \
                 --suite-rep https://github.com/hualongfeng/ceph.git \
                 --suite-branch wip-ioc_cache \
                 --suite-relpath qa \
                 --suite rgw/verify \
                \
                 --ceph-repo https://github.com/hualongfeng/ceph.git \
                 --sha1 bc959f7fbe70f9faed00a970beaf6fce4388f7a9 \
                 --ceph wip-ioc_cache \
                \
                 --kernel distro \
                 --flavor basic \
                \
                 --distro ubuntu \
                 --distro-version 20.04 \
                 --timeout 3600 \
                \
                 --owner hualong  \
                \
                 --filter-out centos

```
 
> where does the `rgw/run-datacache.sh` file come from ?

the `rgw/run-datacache.sh` is from the [script](https://github.com/ceph/ceph/blob/master/qa/workunits/rgw/run-datacache.sh) that you written in D3N PR. It also can use in `rgw-ioc-datachace`.

hualongfeng commented:
@mkogan1  I ran a teuthology test. It can pass. Due to the limited environment, all tasks cannot be run. Can you help test it on teuthology?

[teuthology.log.txt](https://github.com/ceph/ceph/files/7012929/teuthology.log.txt)  This is a log I ran teuthology with rgw-ioc-datacache.
 
Please refer to the command I ran.
```
teuthology-suite -vv  \
                 --machine-type real \
                 --teuthology-branch master \
                 --email hualong.feng@intel.com \
                 \
                 --suite-rep https://github.com/hualongfeng/ceph.git \
                 --suite-branch wip-ioc_cache \
                 --suite-relpath qa \
                 --suite rgw/verify \
                \
                 --ceph-repo https://github.com/hualongfeng/ceph.git \
                 --validate-sha1 false \
                 --sha1 bc959f7fbe70f9faed00a970beaf6fce4388f7a9 \
                \
                 --kernel distro \
                 --flavor basic \
                \
                 --distro ubuntu \
                 --distro-version 20.04 \
                 --timeout 3600 \
                \
                 --owner hualong  \
                \
                 --ceph wip-ioc_cache \
                 --filter-all rgw-ioc-datacache
```

mkogan1 commented:
thank you for the teuthology test, running it.

mkogan1 commented on src/rgw/rgw_ioc_dispatch.h:
```diff
@@ -0,0 +1,142 @@
+                                 RGWObjState *astate, void *arg) {
+  librados::ObjectReadOperation op;
+  struct get_obj_data* d = static_cast<struct get_obj_data*>(arg);
+  string oid, key;
```
compilation err:
```
In file included from ../src/rgw/rgw_aio.cc:22: 
../src/rgw/rgw_ioc_dispatch.h:89:3: error: unknown type name 'string'; did you mean 'std::string'?
  string oid, key;
  ^~~~~~
  std::string
```
```suggestion
  std::string oid, key;
```

hualongfeng replied:
done

mkogan1 commented:
@hualongfeng thank you for rebasing, submitted few teuthology test runs, seems like there is currently a segfault problem during the java s3tests that is not related for IOC, investigating and will re-run tests when possible

hualongfeng commented:
jenkins test make check arm64

hualongfeng commented:
Hi, @mkogan1,  Is there any update about a segfault problem during the java s3tests?

mkogan1 commented:
@hualongfeng no segfaults and teuthology run has completed without unexpected errors.
still have to analyze the teuthology results RGW logs in detail to make sure that the IOC was exercised.

hualongfeng commented:
jenkins test make check arm64

hualongfeng commented:
jenkins test make check arm64

mkogan1 commented:
> Hi, @mkogan1, I fixed the issue. And I checked teuthology log to make sure that the IOC was exercised. Please test it again.

@hualongfeng thank you for rebasing, submited a teuthology run @ https://pulpito.ceph.com/mkogan-2021-11-30_19:40:43-rgw:verify-wip-ioc_cache_i1q-distro-basic-smithi/
care to share a link to your teuthology results and the `teuthology-suite ...` used to submit command please?

hualongfeng commented:
Hi, @mkogan1 , the `teuthology-suite` command is below,
```
#!/bin/bash

teuthology-suite -vv  \
                 --machine-type plana \
                 --teuthology-branch master \
                 --email hualong.feng@intel.com \
                 \
                 --suite-rep https://github.com/hualongfeng/ceph.git \
                 --suite-branch teuthology_ioc \
                 --suite-relpath qa \
                 --suite rgw/verify \
                \
                 --ceph-repo https://github.com/hualongfeng/ceph.git \
                 --validate-sha1 false \
                 --sha1 57fc9cb2b955f2a773da70b0a319f6f52b71ecd2 \
                \
                 --kernel distro \
                 --flavor basic \
                \
                 --distro ubuntu \
                 --distro-version 20.04 \
                 --timeout 3600 \
                \
                 --owner hualong  \
                \
                 --ceph wip-ioc_cache \
                 --filter-out centos,https \
                 --filter-all beast,ignore-pg-availability,msgr-failures/few,objectstore/bluestore-bitmap,overrides,proto/http,rgw_pool_type/ec-profile \

```
About my teuthology results, because it is located in the company's intranet and cannot be accessed from outside, so here I placed a log
[teuthology.log.zip](https://github.com/ceph/ceph/files/7629976/teuthology.log.zip)

mkogan1 commented:
@hualongfeng thank you very much, the teuthology run @ https://pulpito.ceph.com/mkogan-2021-11-30_19:40:43-rgw:verify-wip-ioc_cache_i1q-distro-basic-smithi/ looks good to me, 
checking one of the random IOC runs RGW log for example: http://qa-proxy.ceph.com/teuthology/mkogan-2021-11-30_19:40:43-rgw:verify-wip-ioc_cache_i1q-distro-basic-smithi/6535766/remote/smithi082/log/rgw.ceph.client.0.log.gz
to me seems like IOC is functioning correctly performing reads and writes from cache, 
could you please also review the log and comment whether everything is OK

hualongfeng commented:
Hi, @mkogan1 , I checked the `immutable-object-cache` log 
http://qa-proxy.ceph.com/teuthology/mkogan-2021-11-30_19:40:43-rgw:verify-wip-ioc_cache_i1q-distro-basic-smithi/6535766/remote/smithi082/log/immutable-object-cache.ceph.client.0.log.gz
From the log to see, `immutable-object-cache` daemon is normal to run and communicate with IOC hook.(IOC cache will bypass during writing.)
So we can think everything is OK.

mkogan1 commented:
@hualongfeng great, thank you
submitted full rgw suite (`--suite rgw`) teuthology run:
http://pulpito.ceph.com/mkogan-2021-12-02_09:40:16-rgw-wip-ioc_cache_i11-distro-basic-smithi/

mkogan1 commented:
> @hualongfeng great, thank you submitted full rgw suite (`--suite rgw`) teuthology run: http://pulpito.ceph.com/mkogan-2021-12-02_09:40:16-rgw-wip-ioc_cache_i11-distro-basic-smithi/

none of the failed tests are IOC (or D3N cache) related

mkogan1 reviewed (approved):
approving following passing teuthology test @ https://pulpito.ceph.com/mkogan-2021-12-02_09:40:16-rgw-wip-ioc_cache_i11-distro-basic-smithi/

(the IOC tests are denoted with `datacache/rgw-ioc-datacache`  and the D3N tests are denoted with `datacache/rgw-datacache` in the teuthology tests descriptions)

cbodley commented:
@mkogan1 can we try to get a cleaner run? a recent master baseline at https://pulpito.ceph.com/yuriw-2021-11-25_15:23:03-rgw-wip-yuriw-master-11.24.21-distro-basic-smithi/ only has 4 failures: sts s3-tests failure, tempest test_create_container_with_remove_metadata_value failure, PG_DEGRADED, and multisite

hualongfeng commented:
@mkogan1 Is there any update?

hualongfeng commented:
jenkins test make check

mkogan1 commented:
@cbodley posting updated teuthology results after rebase (did not push the rebase to here yet), seems fine as far as I can see -
https://pulpito.ceph.com/mkogan-2022-03-22_08:45:55-rgw-wip-ioc_cache_n01-distro-default-smithi/
@hualongfeng - if it ok for me to push a rebase here?

hualongfeng commented:
Hi，@mkogan1  @cbodley，is there anything else I need to do here? Thanks.

mkogan1 commented:
> Hi，@mkogan1 @cbodley，is there anything else I need to do here? Thanks.

@hualongfeng following a recent code change, it does not pass compilation after re-basing, may I push a rebase with the fix to this PR?

hualongfeng commented:
> > Hi，@mkogan1 @cbodley，is there anything else I need to do here? Thanks.
> 
> @hualongfeng following a recent code change, it does not pass compilation after re-basing, may I push a rebase with the fix to this PR?
Hi, @mkogan1  Yes, you can. Thanks.

mkogan1 commented:
teuthology run submitted: https://pulpito.ceph.com/mkogan-2022-05-17_08:15:20-rgw-wip-ioc_cache_i17-distro-default-smithi/

hualongfeng commented:
> teuthology run submitted: https://pulpito.ceph.com/mkogan-2022-05-17_08:15:20-rgw-wip-ioc_cache_i17-distro-default-smithi/

Hi, @mkogan1 I don't see the error related to `ioc_cache`. Do we have any next steps? Thanks.

hualongfeng commented:
jenkins retest this

hualongfeng commented:
jenkins retest this please

hualongfeng commented:
Hi, @mkogan1 @cbodley  I rebased the PR.

And I fixed the issue that `g_conf().get_val<size_t>("rgw_max_chunk_size")` will lead **d3n** crash. It should be `g_conf().get_val<Option::size_t>("rgw_max_chunk_size")`.

And now deploy method:

```
cd build

MON=1 OSD=1 MDS=0 MGR=1 RGW=1 ../src/vstart.sh -n -d --bluestore \
           -o "debug immutable_obj_cache=20/20" \
           -o "immutable_object_cache_max_size = 500G" \
           -o "immutable_object_cache_path = /mnt/nvme/ioc_cache/" \
           -o "rgw_d3n_l1_local_datacache_backend=ioc" \
           -o "rgw_d3n_l1_local_datacache_enabled=true" \
           -o "immutable_object_cache_sock=./out/immutable_object_cache_sock.asok" \
           -o rgw_enable_ops_log=true \


ceph-immutable-object-cache -f -c ceph.conf

s3cmd ls s3://testbkt

s3cmd put object s3://testbkt

s3cmd get s3://testbkt/object
```

hualongfeng commented:
jenkins retest

hualongfeng commented:
jenkins test make check arm64

mattbenjamin commented:
*thank you*
